thingif/src/main/java/com/kii/thingif/ThingIFAPI.java

```diff
@@ -1548,8 +1549,30 @@ public String deleteTrigger(
     @NonNull
     @WorkerThread
     public Map<String, ? extends TargetState> getTargetState() throws ThingIFException{
-        //TODO: // FIXME: 12/21/16 implement the logic
-        return new HashMap<>();
+        if (this.target == null) {
+            throw new IllegalStateException("Can not perform this action before onboarding");
+        }
+
+        String path = MessageFormat.format("/thing-if/apps/{0}/targets/{1}/states",
+                this.app.getAppID(), this.target.getTypedID().toString());
+        String url = Path.combine(this.app.getBaseUrl(), path);
+        Map<String, String> headers = this.newHeader();
+        IoTRestRequest request = new IoTRestRequest(url, IoTRestRequest.Method.GET, headers);
+
+        JSONObject responseBody = this.restClient.sendRequest(request);
+        Map retMap = new HashMap<>();
+        Iterator it = responseBody.keys();
+        while (it.hasNext()) {
+            String key = (String)it.next();
+            try {
+                JsonObject object = new JsonParser()
+                        .parse(responseBody.getJSONObject(key).toString()).getAsJsonObject();
+                retMap.put(key, this.gson.fromJson(object, this.stateTypes.get(key)));
+            } catch (JSONException e) {
+                throw new IllegalStateException("Unknown target state class: " + key);
```

Comment:
not throw exception here. Better to check whether key is in this.stateTypes before.